doc/bpf2dagdsm.1
.TH "BPF2DAGDSM" 1 "April 2007"
.SH NAME
bpf2dagdsm \- convert the set of BPF expressions to Endace DSM filter XML specification
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 10
\fBbpf2dagdsm\fR --num-ports=<num-ports> [\fI--max-filters=<number>\fR] [\fI--always-match\fR] <bpf_expressions_file >config.xml
.ad
.hy

.SH "DESCRIPTION"

Note: To fully understand this manual page please read the DAG card documents
first, particularly the one referring to \fBData Stream Management\fR (DSM).

This program converts the Berkeley Packet Filter (\fBBPF\fR) expressions
(well-known as the expressions of \fItcpdump\fR(8)) to \fBXML\fR filter
configuration for Endace \fBDAG\fR cards with the \fBDSM\fR classification.
The resulting XML file can be used as the input for Endace's \fIdsm_loader\fR
executable to set up the hardware filter.

The input is expected on stdin with one complete BPF expression per line
(empty lines are ignored, but counted). Each expression is parsed, compiled
into internal node tree, transformed to Disjunctive Normal Form (DNF, that
is OR-of-ANDs) and broken down to atomic filters.

For each atomic filter one DAG hardware filter is occupied. The resulting
single XML configuration file is printed on stdout and describes all
parsed BPF expressions.

Each group of <filter> elements is followed by the XML <!--comment--> block
containing lines like this one (started with special text tag):

    $DAG-DSM-TAG$ bpf-index:3      filters-used:00110000

The `\fBbpf-index\fR' value is the number of the physical input line
contained the BPF expression. The `\fBfilters-used\fR' is the bit array
describing the association between the BPF expression and hardware filters.
The filters are indexed from zero on the left side. The line above says that
the hardware filters #2 and #3 was allocated for BPF expression on input
line 3. On the next comment line the originating BPF expression is copied
(not shown here).

The line tagged with $DAG-DSM-TAG$ helps the caller to quickly re-analyze
the XML output, construct the filter bitmasks and match triggered filters
back with BPF expressions. To ease the parsing, both values on tagged line
are preceded by the colon (:) and followed by the tab character.

The definition at the bottom of the XML configuration instructs the filter
to pass all packets matching at least one hardware filter to DAG stream 0
(first input stream). Other packets are dropped, but see the option
\fB--always-match\fR below for exception.

.SH "OPTIONS"

.TP
\fB--num-ports=<num-ports>\fR
The number of DAG card ports (interfaces), must be greater than zero.
The XML interface element is generated for each interface. See
\fB--always-match\fR below for how the number is used.

.TP
\fB--max-filters=<number>\fR
When the caller knows the number of hardware filters supported by the DAG card
before calling the \fIbpf2dagdsm\fR, that number can be passed using this option.

When the analyzed configuration exceeds this number, no XML output is printed.
Instead the error message is sent do stderr and return code is nonzero.

This is the safety measure, because the \fIdsm_loader\fR blindly loads
everything until all hardware filters are filled and omits the rest.
Then this damaged DAG card configuration is activated. This option allows for
detection of such error and skipping the bogus call of \fIdsm_loader\fR.

.TP
\fB--always-match\fR
This options merely adds one OR'ed partial element for each existing interface.

As result, all packets (even those not matching the previous filters) are
also accepted (not filtered out by the DAG card). The triggered filters can
still be distinguished by analyzing the ERF header. Using this option, the
DAG card stops filtering and becomes the packet classification machine.

.SH "RESTRICTIONS"

Unfortunately, only a subset of BPF expression features is implemented. This
is mainly due to the fact that the original BPF is fed to a machine,
processing the true algorithm for each examined packet. On the other hand,
DSM-driven hardware can only match static bitmasked filters. So all BPF
constructs that require "runtime" computation must have been omited.

Also currently the program is capable of generating only DAG filters using
symbolic XML elements (called "virtual" by Endace), not "raw" filters
containing direct maskable true/false values for every bit in the first 64
bytes of the packets captured. Therefore in the present time only the source
or destination IPv4 addresses of Ethernet frames can be matched. Additionally,
filters can specify TCP or UDP source or destination ports.

.SH "RETURN VALUE"

.TP
0
Entire XML filter successfully constructed and printed out.

.TP
!= 0
Error occured or help text printed. The problem is described on stderr.

.SH TODO

The program should be able to re-use duplicated filters, probably by comparing
their string representation generated by function print_node().



debian/changelog
mapi (2.0.uninett.20120110-1) unstable; urgency=low

  * Fixed premature free which broke offline processing. svn version 1608.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 10 Jan 2012 10:53:43 +0100

mapi (2.0.uninett.20100504-2) unstable; urgency=low

  * Read minFlowPkts for ipfixlib from config file. Existing min size logic in bytes changed to packets.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 04 May 2010 14:44:33 +0200

mapi (2.0.uninett.20100504-1) unstable; urgency=low

  * Fix another race in ipfixlib.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 04 May 2010 13:25:46 +0200

mapi (2.0.uninett.20100429-1) unstable; urgency=low

  * Fix race in ipfixlib. svn version 1570.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu, 29 Apr 2010 10:23:15 +0200

mapi (2.0.uninett.20100428-1) unstable; urgency=low

  * Load balancing tricks for napatech.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Wed, 28 Apr 2010 11:32:51 +0200

mapi (2.0.uninett.20100326-1) unstable; urgency=low

  * MPEG-TS completed.
  svn version 1567.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Fri, 26 Mar 2010 13:25:25 +0100

mapi (2.0.uninett.20100212-1) unstable; urgency=low

  * Fix a few qflow rtp problems.
  svn version 1558.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Fri, 12 Feb 2010 14:36:21 +0100

mapi (2.0.uninett.20100209-1) unstable; urgency=low

  * Slight change in qflow field names.
  svn version 1557.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 09 Feb 2010 16:10:03 +0100

mapi (2.0.uninett.20091202-1) unstable; urgency=low

  * Fix segfault in sip_parse in ipfixlib. 
  svn version 1555.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Wed, 02 Dec 2009 09:44:59 +0100

mapi (2.0.uninett.20091124-1) unstable; urgency=low

  * Buil for lenny.
  svn version 1554.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 24 Nov 2009 12:02:42 +0100

mapi (2.0.uninett.20091106-1) unstable; urgency=low

  * Turn off a little debug output.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Fri,  6 Nov 2009 13:52:25 +0100

mapi (2.0.uninett.20091105-1) unstable; urgency=low

  * Include Napatech NT in build. Remove appmon.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu,  5 Nov 2009 10:50:23 +0100

mapi (2.0.uninett.20090924-1) unstable; urgency=low

  * Disable mpeg ts unless ENABLE_MPEGTS is on.
  svn version 1539.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu, 24 Sep 2009 12:55:14 +0200

mapi (2.0.uninett.20090907-1) unstable; urgency=low

  * Don't try to string match services when there's no payload.
  svn version 1537.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Mon,  7 Sep 2009 14:37:15 +0200

mapi (2.0.uninett.20090817-1) unstable; urgency=low

  * More mpeg ts fixes.
  svn version 1535.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Mon, 17 Aug 2009 15:48:40 +0200

mapi (2.0.uninett.20090814-1) unstable; urgency=low

  * More mpeg ts fixes.
  svn version 1534.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Fri, 14 Aug 2009 10:51:02 +0200

mapi (2.0.uninett.20090813-1) unstable; urgency=low

  * Tweak mpeg ts recognition.
  svn version 1533.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu, 13 Aug 2009 13:22:20 +0200

mapi (2.0.uninett.20090812-1) unstable; urgency=low

  * ipfixlib: Add mpeg support.
  svn version 1532.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 11 Aug 2009 15:51:09 +0200

mapi (2.0.uninett.20090324-1) unstable; urgency=low

  * mapidump: Add options to print ifindex and to capture from a particular 
  ifindex.
  svn version 1489.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue, 24 Mar 2009 12:58:02 +0100

mapi (2.0.uninett.20090319-1) unstable; urgency=low

  * Ipfixlib: Report true ingress and egress interface. Only works for 
  passive monitoring cards which see both directions. 
  svn version 1486.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu, 19 Mar 2009 09:52:17 +0100

mapi (2.0.uninett.20090204-1) unstable; urgency=low

  * More appmon postinst tweaks.
  svn version 1473.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Wed,  4 Feb 2009 11:23:49 +0100

mapi (2.0.uninett.20090203-2) unstable; urgency=low

  * Add back appmon prerm and postinst, and make them more robust. Remember to
  add them to svn.
  svn version 1472.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue,  3 Feb 2009 17:01:12 +0100

mapi (2.0.uninett.20090203-1) unstable; urgency=low

  * Plug memory leak in services.c.
  * Clean up debug ifdefs.
  * Harmonize traceEvent and traceMode with new upstream nprobe.    
  * Add traceMode, traceToFile and statInterval parameters, only call
    printStat once every statInterval scan.
  * Possible to trace to file instead of stdout.
  * Trace avg bucket search length.
  svn version 1469.  

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Tue,  3 Feb 2009 14:45:08 +0100

mapi (2.0.uninett.20090122-1) unstable; urgency=low

  * Fix broken build.
  * Plug a few minor leaks at flow probe exit.
  svn version 1456.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Thu, 22 Jan 2009 14:18:53 +0100

mapi (2.0.uninett.20090120-1) unstable; urgency=low

  * Make scanCycle and hashSize config parameters.
  svn version 1453. 
  
 --  Jon K Hellan <jon.kare.hellan@uninett.no>  Tue 20 Jan 2009 13:41:59 +0100

mapi (2.0.uninett.20090119-4) unstable; urgency=low

  * Fix leakage of cached flow buckets.
  svn version 1452.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Mon, 19 Jan 2009 15:49:49 +0100

mapi (2.0.uninett.20090119-3) unstable; urgency=low

  * Remove cache debugging output. 
  * Replace some #if 0/#if 1 with meaningful #ifdef DEBUG_FOO switches. 
  svn version 1451.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Mon, 19 Jan 2009 14:35:51 +0100

mapi (2.0.uninett.20090119-2) unstable; urgency=low

  * Use only 1 mutex. Flow caching code doesn't work with more than 1.
  * Increase no of hash buckets to 65536.

 -- Jon K Hellan <jon.kare.hellan@uninett.no>  Mon, 19 Jan 2009 14:11:23 +0100


src/dagflib/bpfcompiler/bpf2dagdsm.c
/*
 *
 * LICENSE TERMS
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *  3. Neither the name of the Company nor the names of its contributors
 *     may be used to endorse or promote products derived from this
 *     software without specific prior written permission.
 *
 * ALTERNATIVELY, provided that this notice is retained in full, this
 * product may be distributed under the terms of the GNU General Public
 * License (GPL) version 2 or later, in which case the provisions
 * of the GPL apply INSTEAD OF those given above.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'', AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COMPANY OR CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <limits.h>
#include <stdarg.h>

#include "utils.h"
#include "bpfcompiler.h"

#define FATAL(str, rest...) fatal("internal error in %s:%d: " str "\n", __FILE__, __LINE__, ## rest)

void fatal(const char *fmt, ...)
        __attribute__((noreturn, format (printf, 1, 2)));

void fatal(const char *fmt, ...)
{
    va_list ap;

    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
    exit(-1);
}

void usage(char *progname)
{
    fprintf(stderr, "Usage: %s --num-ports=<num-ports> [--max-filters=<number>] [--always-match]\n", progname);    
    exit(100);
}

void parse_args(int argc, char **argv, int *max_filters, int *always_match, int *num_ports)
{
    int i;
    *always_match = 0;
    *max_filters = INT_MAX;
    *num_ports = 0;

    for (i=1; i<argc; i++) {
    	if (!strncmp("--num-ports=", argv[i], 12)) {
            *num_ports = atoi(argv[i] + 12);
    	} else if (!strncmp("--max-filters=", argv[i], 14)) {
            *max_filters = atoi(argv[i] + 14);
            if (*max_filters < 1)
                usage(argv[0]);
        } else if (!strcmp("--always-match", argv[i])) {
            (*always_match)++;
        } else if (!strcmp("--help", argv[i]) || !strcmp("-h", argv[i])) {
            usage(argv[0]);
        } else
            fatal("%s: Unknown or misformatted parameter '%s'. Use --help for help.\n", argv[0], argv[i]);
    }
    if (*num_ports <= 0)
    	usage(argv[0]);
}

int main(int argc, char **argv)
{
    char errbuf[BPF_ERRBUF_SIZE], inbuf[4096], *output;
    int lineno = 0, max_filters, always_match, filter_count, num_ports;
    dagdsm_bpf_filter_t *filter = NULL;

    parse_args(argc, argv, &max_filters, &always_match, &num_ports); 

    while (fgets(inbuf, sizeof(inbuf), stdin)) {
        char *p = strchr(inbuf, '\n'); 

        lineno++;
        if (!p)
            fatal("%s: Fatal error, the length of input line %d is above %u bytes.\n", argv[0], lineno, sizeof(inbuf)-1);
        
        *p = '\0';
        if (inbuf[0] == '\0')
            continue;

        /* Compile BPF filter */
        if ((filter = dagdsm_bpf_compile(filter, inbuf, lineno, errbuf)) == NULL)
            fatal("%s: Compilation error on input line %d: %s\n", argv[0], lineno, errbuf);

        BPF_DEBUG_CMD(fprintf(stderr, "line %d filtermask 0x%08x\n", lineno, dagdsm_bpf_get_last_usage_mask(filter)));
    }

    filter_count = dagdsm_bpf_get_filter_count(filter);
    
    if (filter_count > max_filters)
        fatal("%s: The number of filters needed (%d) exceeds the given limit (%d).\n", argv[0], filter_count, max_filters);

    /* Write the filter to DAGDSM configuration output file. */
    output = dagdsm_bpf_xprintf(filter, always_match, num_ports, errbuf);

    /* Free resources allocated for the filter. */
    if (filter != NULL)
        dagdsm_bpf_free(filter);

    if (output) 
    	printf(output);
    else
    	fatal("%s: Compilation error on input line %d: %s\n", argv[0], lineno, errbuf);

    free(output);

    return EXIT_SUCCESS;
}
